### ISSUE TITLE
ClientContext causing AddressSanitizer error: requested allocation size exceeds maximum supported size

### ISSUE THREAD

**Issue by aurelienrb** [grpc/grpc#21838]
### What version of gRPC and what language are you using?

Tested with gRPC v1.25.0 and 1.23.1

### What operating system (Linux, Windows,...) and version?

Linux

### What runtime / compiler are you using (e.g. python version or version of gcc)

g++8 with `-fsanitize=address`

### What did you do?

We use gRPC to run various C++ servers, and from time to time they tend to crash during their startup phase. So I built a version with AddressSanitizer enabled, and it reports again and again the same issue in `grpc::internal::CallOpSendInitialMetadata::AddOp(grpc_op*, unsigned long*) /usr/local/grpc/include/grpcpp/impl/codegen/call_op_set.h:243`

```
ERROR: AddressSanitizer: requested allocation size 0x33c5312812c800 (0x33c5312812d800 after adjustments for alignment, red zones etc.) exceeds maximum supported size of 0x10000000000 (thread T28)

    #0 0x7f6f6b6f9f88 in malloc (/usr/lib/x86_64-linux-gnu/libasan.so.5+0x10bf88)
    #1 0xc2816e in gpr_malloc (/tabmo/app/bidder/build/bin/router_service+0xc2816e)
    #2 0x5652e9 in grpc::internal::CallOpSendInitialMetadata::AddOp(grpc_op*, unsigned long*) /usr/local/grpc/include/grpcpp/impl/codegen/call_op_set.h:243
    #3 0x57f948 in grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal::CallOpSendMessage, grpc::internal::CallOpRecvInitialMetadata, grpc::internal::CallOpRecvMessage<bidder::agentservice::StatusResponse>, grpc::internal::CallOpClientSendClose, grpc::internal::CallOpClientRecvStatus>::ContinueFillOpsAfterInterception() (/tabmo/app/bidder/build/bin/router_service+0x57f948)
    #4 0x57f737 in grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal::CallOpSendMessage, grpc::internal::CallOpRecvInitialMetadata, grpc::internal::CallOpRecvMessage<bidder::agentservice::StatusResponse>, grpc::internal::CallOpClientSendClose, grpc::internal::CallOpClientRecvStatus>::FillOps(grpc::internal::Call*) (/tabmo/app/bidder/build/bin/router_service+0x57f737)
    #5 0x55d075 in grpc::internal::Call::PerformOps(grpc::internal::CallOpSetInterface*) /usr/local/grpc/include/grpcpp/impl/codegen/call.h:69
    #6 0x573fc9 in grpc::internal::BlockingUnaryCallImpl<bidder::agentservice::StatusRequest, bidder::agentservice::StatusResponse>::BlockingUnaryCallImpl(grpc::ChannelInterface*, grpc::internal::RpcMethod const&, grpc_impl::ClientContext*, bidder::agentservice::StatusRequest const&, bidder::agentservice::StatusResponse*) (/tabmo/app/bidder/build/bin/router_service+0x573fc9)
    #7 0x56ff2b in grpc::Status grpc::internal::BlockingUnaryCall<bidder::agentservice::StatusRequest, bidder::agentservice::StatusResponse>(grpc::ChannelInterface*, grpc::internal::RpcMethod const&, grpc_impl::ClientContext*, bidder::agentservice::StatusRequest const&, bidder::agentservice::StatusResponse*) /usr/local/grpc/include/grpcpp/impl/codegen/client_unary_call.h:43
    #8 0x55bf09 in bidder::agentservice::MasterConfigPool::Stub::Status(grpc_impl::ClientContext*, bidder::agentservice::StatusRequest const&, bidder::agentservice::StatusResponse*) /builds/data-science/bidder/cmake-build/rtbkit/master_grpc_common/agent_service.grpc.pb.cc:37
    #9 0x54cc32 in Bidder::Client::MasterConfigPoolGrpcClient::status(std::chrono::duration<long, std::ratio<1l, 1000l> >) /builds/data-science/bidder/rtbkit/master_grpc_common/agent_service_client.cc:34
    #10 0x4b8c85 in Router::RouterService::queryMasterPoolVersion() /builds/data-science/bidder/rtbkit/core/router/router.cc:383
```

### What did you see instead?

- if I disable the gRPC calls, our software runs well
- if I disable the first grpc service being called, it crashed on the next one (same issue)
- it tends to crash during the first call to the grpc service
- if double checked (with tsan) for possible memory corruption on our side, but I don't see how it can happen for this trivial part

### Anything else we should know about your project / environment?

I though it migh be caused by something wrong on our side, until I spotted that **TODO** comment in the grpc code where the crash happens:

> TODO(yangg) if the map is changed before we send, the pointers will be a
> mess. Make sure it does not happen.

```cpp
// TODO(yangg) if the map is changed before we send, the pointers will be a
// mess. Make sure it does not happen.
inline grpc_metadata* FillMetadataArray(
    const std::multimap<grpc::string, grpc::string>& metadata,
    size_t* metadata_count, const grpc::string& optional_error_details) {
  *metadata_count = metadata.size() + (optional_error_details.empty() ? 0 : 1);
  if (*metadata_count == 0) {
    return nullptr;
  }
  grpc_metadata* metadata_array =
      (grpc_metadata*)(g_core_codegen_interface->gpr_malloc(
          (*metadata_count) * sizeof(grpc_metadata)));
  size_t i = 0;
  for (auto iter = metadata.cbegin(); iter != metadata.cend(); ++iter, ++i) {
    metadata_array[i].key = SliceReferencingString(iter->first);
    metadata_array[i].value = SliceReferencingString(iter->second);
  }
  if (!optional_error_details.empty()) {
    metadata_array[i].key =
        g_core_codegen_interface->grpc_slice_from_static_buffer(
            kBinaryErrorDetailsKey, sizeof(kBinaryErrorDetailsKey) - 1);
    metadata_array[i].value = SliceReferencingString(optional_error_details);
  }
  return metadata_array;
}
```

this is where it is called in grpc code:

```cpp
  void AddOp(grpc_op* ops, size_t* nops) {
    if (!send_ || hijacked_) return;
    grpc_op* op = &ops[(*nops)++];
    op->op = GRPC_OP_SEND_INITIAL_METADATA;
    op->flags = flags_;
    op->reserved = NULL;
    initial_metadata_ =
        FillMetadataArray(*metadata_map_, &initial_metadata_count_, "");
```

Do you have more details about that comment and what it implies ? Could it be the cause of our issue ?

**Comment by aurelienrb** [comment]
Reading more closely the source code, the issue might be caused by the combination of the following grpc code that initialize the map being referenced later where the crash happens :

```cpp
  void ClientRecvStatus(::grpc_impl::ClientContext* context, Status* status) {
    client_context_ = context;
    metadata_map_ = &client_context_->trailing_metadata_;
    recv_status_ = status;
    error_message_ = g_core_codegen_interface->grpc_empty_slice();
  }
```
and our client code that looks like this:
```cpp
std::optional<MasterPoolStatus> MasterConfigPoolGrpcClient::status(std::chrono::milliseconds timeout) {
    grpc::ClientContext ctx;
    setContextTimeout(ctx, timeout);

    bidder::agentservice::StatusResponse response;
    bidder::agentservice::StatusRequest request;
    const auto status = m_pimpl->stub->Status(&ctx, request, &response);
    if (status.ok()) {
        MasterPoolStatus result;
        result.poolVersion = response.poolversion();
        result.configCount = response.configcount();
        return result;
    }
    return {};
}
```

indeed, I managed to get my compiler crash once in this function, when calling the destructor of the local `ClientContext`.

So the question is : is it safe to immediately destroy the `ClientContext` that was used for a synchronous call once that call has returned?

**Comment by aurelienrb** [comment]
I think I found the reason why it breaks with ASAN: it's a good old violation of the ODR. Here's the reason why:

`grpc::ClientContext` embeds a `grpc::internal::Mutex` which embeds `gpr_mu` which is defined like this:

```cpp
#ifdef GRPC_ASAN_ENABLED
/* The member |leak_checker| is used to check whether there is a memory leak
 * caused by upper layer logic that's missing the |gpr_xx_destroy| call
 * to the object before freeing it.
 * This issue was reported at https://github.com/grpc/grpc/issues/17563
 * and discussed at https://github.com/grpc/grpc/pull/17586
 */
typedef struct {
  pthread_mutex_t mutex;
  int* leak_checker;
} gpr_mu;

typedef struct {
  pthread_cond_t cond_var;
  int* leak_checker;
} gpr_cv;
#else
typedef pthread_mutex_t gpr_mu;
typedef pthread_cond_t gpr_cv;
#endif
```

therefore, if the client code is built with ASAN but GRPC is not, then :boom:

**Comment by karthikravis** [comment]
Hey @veblush , I remember you working on ASAN/TSAN and gRPC builds.. can you please take a look and advise?

**Comment by bitomaxsp** [comment]
Hi. We kind of blocked by this issue. Is there any progress here?
Was it fixed?

**Comment by aurelienrb** [comment]
Hello,
I have something to propose that I tested with success locally. The idea it to cause a linker error if you try to link with a gRPC lib that was not built with same options than yours (for the settings that matters, i.e. that can cause ODR violation). Its totally transparent for the client code.
What do you think about this approach? Would you like me to submit a PR for comments?

**Comment by bitomaxsp** [comment]
I think linker error is better then debugging ODR violation :)
It would be nice to have PR submitted for the comments.

**Comment by Mekk** [comment]
Let me say that it's just crazy. For „normal human” libgrpc.so etc are `apt install`-ed, or mayhaps compiled by admin from time to time. The idea that using sanitizer in my code should force me to recompile grpc is … well, not really appealing.

Fortunately there exist relatively easy workaround:
```
#include <grpc/impl/codegen/port_platform.h>
#ifdef GRPC_TSAN_ENABLED
#  undef GRPC_TSAN_ENABLED
#endif
#ifdef GRPC_ASAN_ENABLED
#  undef GRPC_ASAN_ENABLED
#endif
```
(I add it before any other grpc workaround, and scriptlety-prepend to grpc-generated .cpp files)

**Comment by pyssling** [comment]
@Mekk good idea this. I took it one-step further, I removed this code from the port_platform.h header file on my system - I think I'll file a packaging bug and try to get this removed from packages of the grpc headers. If grpc isn't built with ASAN there's no point in shipping this in a header.

**Comment by pyssling** [comment]
@veblush I think this would be best to fix by stripping the headers for non-asan builds of grpc of these ifdefs - the same goes for non-debug builds of grpc and the NDEBUG ifdefs - breaking ABI compatibility of library users creates a lot of extra work.

**Comment by mortie** [comment]
Hi, this is still a problem. Could the issue be re-opened?

**Comment by Mekk** [comment]
Not sure who can reopen (I can't).

I still use workaround described above.
